proc-macros/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+#[proc_macro_attribute]
+pub fn topic(args: TokenStream, input: TokenStream) -> TokenStream {
+    let new_input = proc_macro2::TokenStream::from(input);
+    let mut iter = new_input.clone().into_iter().filter(|i| match i {
+        TokenTree::Group(_) => true,
+        TokenTree::Ident(_) => true,
+        TokenTree::Punct(_) => false,
+        TokenTree::Literal(_) => false,
+    });
+
+    let mut current = iter.next().unwrap();
+
+    while current.to_string() != "fn" {
+        current = iter.next().unwrap()
+    }
+
+    let name = iter.next().unwrap();
+
+    let struct_name = name
+        .to_string()
+        .split('_')
+        .into_iter()
+        .map(|i| {
+            let mut chars: Vec<char> = i.chars().collect();
+            chars[0] = chars[0].to_ascii_uppercase();
+            let new_string: String = chars.into_iter().collect();
+            new_string
+        })
+        .collect::<Vec<String>>()
+        .join("");
+
+    let name_ident = Ident::new(name.to_string().as_str(), name.span());
+
+    let struct_name_ident = Ident::new(struct_name.as_str(), name.span());
+
+    let vars: Vec<String> = iter
+        .next()
+        .unwrap()
+        .to_string()
+        .replace(['(', ')'], "")
+        .split(':')
+        .into_iter()
+        .enumerate()
+        .filter(|&(i, _)| i % 2 != 0)
+        .map(|(_, i)| i.trim().to_string())
+        .collect();
+
+    assert_eq!(vars.len(), 1, "Expected to only have one input variable");
+
+    let parse_statement = match vars[0] == *"String" {
+        true => {
+            quote! {
+                let message = message.to_string();
+            }
+        }
+        false => {
+            let type_ident = format_ident!("{}", vars[0]);
+            println!("{}", type_ident);
+            quote! {
+               let message: #type_ident = dapr::serde_json::from_str(message.to_string().as_str()).unwrap();
+            }
+        }
+    };
+
+    let args = parse_macro_input!(args as TopicArgs);
+
+    let topic = args.topic.value();
+
+    let pub_sub_name = args.pub_sub_name.value();
+
+    let tokens = quote! {
+        #new_input
+
+        #[derive(Default)]
+        struct #struct_name_ident;
+
+        #[tonic::async_trait]
+        impl AppCallback for #struct_name_ident {
+            async fn on_invoke(
+                &self,
+                _request: Request<InvokeRequest>,
+            ) -> Result<Response<InvokeResponse>, Status> {
+                Ok(Response::new(InvokeResponse::default()))
+            }
+
+            async fn list_topic_subscriptions(
+                &self,
+                _request: Request<()>,
+            ) -> Result<Response<ListTopicSubscriptionsResponse>, Status> {
+                let topic = #topic.to_string();
+                let pubsub_name = #pub_sub_name.to_string();
+
+                let list_subscriptions = ListTopicSubscriptionsResponse::topic(pubsub_name, topic);
```

Comment:
It looks like this would only work for a single topic.

The `ListTopicSubscriptionsResponse::topic` method is a bit misleading. See https://github.com/dapr/rust-sdk/issues/81#issuecomment-1264094970 to see how to add multiple topics.

Reply:
Oh, my bad, I see each use of the macro creates an AppCallback implementation.

As far as I could see, this doesn't work with tonic (see my other comment here: https://github.com/dapr/rust-sdk/issues/81#issuecomment-1264116605).

So I think we need to have a single implementation generated, but I don't know how that can work in this case.

Could something like this work?

```rs
struct MyService {}

#[dapr] // with this creating the single AppCallback implementation
impl MyService {

  // The pubsub_name and topic would be stored, and the method would then be used in the   `on_topic_event` trait implementation (eg: using match).
  // perhaps the function signature should have a parameter for pubsub name and topic (eg a Context struct)
  #[topic(pubsub_name = "pubsub", topic = "A")]
  async fn handle_message(&self, order: Order) {
    println!("{:#?}", order)
  }

}
```

Keen to hear your thoughts?